go/vt/vtgate/healthcheck.go

```diff
@@ -0,0 +1,279 @@
+package vtgate
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/golang/glog"
+	pbq "github.com/youtube/vitess/go/vt/proto/query"
+	pbt "github.com/youtube/vitess/go/vt/proto/topodata"
+	"github.com/youtube/vitess/go/vt/tabletserver/tabletconn"
+	"github.com/youtube/vitess/go/vt/topo"
+	"golang.org/x/net/context"
+)
+
+// AddrUpdate updates about an endpoint and whether it is up.
+type AddrUpdate struct {
+	EndPoint *pbt.EndPoint
+	IsUp     bool
+}
+
+// NewHealthCheck creates a new HealthCheck object.
+func NewHealthCheck(endPoints []*pbt.EndPoint, addrUpdate <-chan *AddrUpdate, connTimeout time.Duration, retryDelay time.Duration) *HealthCheck {
+	hc := &HealthCheck{
+		addrToConns: make(map[string]*healthCheckConn),
+		targetToEPs: make(map[string]map[string]map[pbt.TabletType][]*pbt.EndPoint),
+		connTimeout: connTimeout,
+		retryDelay:  retryDelay,
+	}
+	hc.initConns(endPoints)
+	go hc.onAddrUpdate(addrUpdate)
+	return hc
+}
+
+// HealthCheck performs health checking and notifies downstream components about any changes.
+type HealthCheck struct {
+	mu          sync.RWMutex
+	addrToConns map[string]*healthCheckConn                              // addrToConns maps from address to the healthCheckConn object.
+	targetToEPs map[string]map[string]map[pbt.TabletType][]*pbt.EndPoint // targetToEPs maps from keyspace/shard/tablettype to a list of endpoints.
+	connTimeout time.Duration
+	retryDelay  time.Duration
+}
+
+// healthCheckConn contains details about an endpoint.
+type healthCheckConn struct {
+	mu         sync.RWMutex
+	target     *pbq.Target
+	cancelFunc context.CancelFunc
+	conn       tabletconn.TabletConn
+	stats      *pbq.RealtimeStats
+}
+
+// initConns starts health checking for the initial set of endpoints.
+// It waits till all endpoints are tried to connect once.
+func (hc *HealthCheck) initConns(endPoints []*pbt.EndPoint) {
+	var wg sync.WaitGroup
+	for _, endPoint := range endPoints {
+		wg.Add(1)
+		go hc.checkConn(endPoint, &wg)
+	}
+	wg.Wait()
+}
+
+// checkConn performs health checking on the given endpoint.
+func (hc *HealthCheck) checkConn(endPoint *pbt.EndPoint, wg *sync.WaitGroup) {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	hcc := &healthCheckConn{
+		cancelFunc: cancelFunc,
+	}
+	addr := hc.endPointToString(endPoint)
+
+	defer func() {
+		// Clean up the healthCheckConn from maps, and notify addrwatcher
+		hc.mu.Lock()
+		// TODO: notify downstream component for endpoint going down
+		if hcc.target != nil {
+			hc.deleteEndPointFromTarget(hcc.target, endPoint)
+			delete(hc.addrToConns, addr)
+		}
+		hc.mu.Unlock()
+	}()
+
+	var conn tabletconn.TabletConn
+	var stream <-chan *pbq.StreamHealthResponse
+	var errfunc tabletconn.ErrFunc
+	var err error
+
+	// func to read one health check response, and notify downstream component
+	processResponse := func() error {
```

Comment:
I'm not a big fan of these closures, could these two just be methods of the hcc? I think the code would be clearer then, and easier to unit test HCC by itself.